File: backend/app/schemas/comptabilite.py
from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class CategorieDepenseResponse(BaseModel):
    id: UUID
    code: str
    libelle: str
    actif: bool

    model_config = {"from_attributes": True}


class CompteTresorerieResponse(BaseModel):
    id: UUID
    code: str
    libelle: str
    type: str
    solde_initial: Decimal
    solde_actuel: Decimal | None = None
    actif: bool

    model_config = {"from_attributes": True}


class DepenseCreate(BaseModel):
    categorie_id: UUID
    annee_scolaire_id: UUID | None = None
    libelle: str = Field(..., min_length=3)
    montant: Decimal = Field(..., gt=0)
    date_depense: date
    compte_tresorerie_id: UUID
    reference_piece: str | None = None


class DepenseRefus(BaseModel):
    motif: str = Field(..., min_length=3)


class DepenseResponse(BaseModel):
    id: UUID
    categorie_id: UUID
    categorie_libelle: str
    annee_scolaire_id: UUID
    libelle: str
    montant: Decimal
    date_depense: date
    compte_tresorerie_id: UUID
    compte_libelle: str
    statut: str
    reference_piece: str | None
    motif_refus: str | None


class BudgetLigneCreate(BaseModel):
    annee_scolaire_id: UUID
    categorie_id: UUID
    montant_prevu: Decimal = Field(..., ge=0)


class BudgetSuiviItem(BaseModel):
    categorie_id: UUID
    categorie_code: str
    categorie_libelle: str
    montant_prevu: Decimal
    montant_realise: Decimal
    ecart: Decimal
    taux_realisation: Decimal | None


class BudgetSuiviResponse(BaseModel):
    annee_scolaire_id: UUID
    annee_libelle: str
    total_prevu: Decimal
    total_realise: Decimal
    lignes: list[BudgetSuiviItem]


class EcritureResponse(BaseModel):
    id: UUID
    date_ecriture: date
    type: str
    libelle: str
    montant: Decimal
    compte_libelle: str
    source_type: str | None


class RapportFinancierResponse(BaseModel):
    periode_debut: date
    periode_fin: date
    total_recettes: Decimal
    total_depenses: Decimal
    solde: Decimal
    recettes_par_mode: dict[str, Decimal]
    depenses_par_categorie: dict[str, Decimal]


class TresorerieResponse(BaseModel):
    comptes: list[CompteTresorerieResponse]
    total_caisse: Decimal
    total_banque: Decimal
    total_general: Decimal

File: frontend/src/types/comptabilite.ts
export type CategorieDepense = {
  id: string;
  code: string;
  libelle: string;
  actif: boolean;
};

export type CompteTresorerie = {
  id: string;
  code: string;
  libelle: string;
  type: string;
  solde_initial: string;
  solde_actuel?: string;
  actif: boolean;
};

export type Depense = {
  id: string;
  categorie_id: string;
  categorie_libelle: string;
  annee_scolaire_id: string;
  libelle: string;
  montant: string;
  date_depense: string;
  compte_tresorerie_id: string;
  compte_libelle: string;
  statut: string;
  reference_piece: string | null;
  motif_refus: string | null;
};

export type BudgetSuiviItem = {
  categorie_id: string;
  categorie_code: string;
  categorie_libelle: string;
  montant_prevu: string;
  montant_realise: string;
  ecart: string;
  taux_realisation: string | null;
};

export type BudgetSuivi = {
  annee_scolaire_id: string;
  annee_libelle: string;
  total_prevu: string;
  total_realise: string;
  lignes: BudgetSuiviItem[];
};

export type Ecriture = {
  id: string;
  date_ecriture: string;
  type: string;
  libelle: string;
  montant: string;
  compte_libelle: string;
  source_type: string | null;
};

export type RapportFinancier = {
  periode_debut: string;
  periode_fin: string;
  total_recettes: string;
  total_depenses: string;
  solde: string;
  recettes_par_mode: Record<string, string>;
  depenses_par_categorie: Record<string, string>;
};

export type Tresorerie = {
  comptes: CompteTresorerie[];
  total_caisse: string;
  total_banque: string;
  total_general: string;
};
